README.md
# camunda-external-microservice
Uses external task client to redirect variable microservices to modules

## microApi

Acts as an interface between Camunda and other microservices and executes simple tasks.
Implemented using the External Task Client from Camunda (https://github.com/camunda/camunda-external-task-client-js).

### Configuration

Default method for setting configurations options is the .env file.

If you are using PM2 as process manager, you can use their ecosystem json file declaration (sample
in app.json.sample). In case of PM2 delete the .env file, because it will overwrite the variables
of PM2.

### Option

| Variable | Type | Optional | Default | Values | Description |
|---|---|---|---|---|---|
| NODE_ENV | String |  | -  | development; staging; production  | node environment
|  TOPIC |  String |   | -  |   | external service task topic |
| ASYNC_RESPONSE_TIMEOUT |  Number | X  | -  |   | [External Task Client Docs](https://github.com/camunda/camunda-external-task-client-js/blob/master/docs/Client.md#new-clientoptions)
| MAX_TASK |  Number | X  | 10  |   | [External Task Client Docs](https://github.com/camunda/camunda-external-task-client-js/blob/master/docs/Client.md#new-clientoptions)
| INTERVAL |  Number | X  | 300  |   | [External Task Client Docs](https://github.com/camunda/camunda-external-task-client-js/blob/master/docs/Client.md#new-clientoptions)
| LOCK_DURATION |  Number | X  | 50000  |   | [External Task Client Docs](https://github.com/camunda/camunda-external-task-client-js/blob/master/docs/Client.md#new-clientoptions)
| BASE_URL |  String | X  | http://localhost:8080/engine-rest  |   | [External Task Client Docs](https://github.com/camunda/camunda-external-task-client-js/blob/master/docs/Client.md#new-clientoptions)
| WORKER_ID |  String | X  | default-worker  |   | [External Task Client Docs](https://github.com/camunda/camunda-external-task-client-js/blob/master/docs/Client.md#new-clientoptions)
| AUTH |  Boolean | X  | false  |   | Use basic auth for REST
| USER |  String | X  | -  |   | Basic auth user
| PASS |  String | X  | -  |   | Basic auth password

### How it works

Place a service task in your BPMN and configure it to use the external microservice (see the sample BPMN).

This service will grab the task and execute the specified module.  
The module is specified as input parameter named "micro-api-module" and as value the filename of the
the module (without extension).

You can set more input variables in the service task to use them directly in the module. Modules
then take care of some service (e.g. Query another API, build a PDF, download stuff etc.).

### Example

Sample folder holds a BPMN using the "just-complete"-Microservice.
It just completes the service successfully.

The http-request calls a mock server for some JSON and sets a process variable.

### Concept

The concept differs in the way the Camunda external task client works by using
the variable module name instead of using a variable topic name.

Bernd Rücker talks about this kind of concept and wether it's a good or bad practice.

A discussion about it can be found in the Camunda forum:  
https://forum.camunda.org/t/microservices-external-tasks-not-the-best-practice/15387

### LICENCE

MIT License

Copyright (c), 2021, KAJU.IT UG (haftungsbeschränkt)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
#Changelog

## 1.1.0 (28.09.2021)

- optional logging
- .env file for environment variables

## 1.0.0 (21.09.2021)

- Initial release


log.ts
import * as winston from 'winston';
import 'winston-daily-rotate-file';
import * as path from "path";

export let log;

export function creatLogger() {
    let logPath = path.join(__dirname, 'logs');
    let transports = [];

    if (process.env.FILE_LOG !== 'false') {
        const transport = new (winston.transports.DailyRotateFile)({
            filename: 'application-%DATE%.log',
            dirname: logPath,
            datePattern: 'YYYY-MM-DD-HH',
            zippedArchive: true,
            maxSize: '15m',
            maxFiles: '7d'
        });

        // transport.on('rotate', function(oldFilename, newFilename) {
        // do something fun
        // });

        transports.push(transport);
    }

    var winstonLogger = winston.createLogger({
        level: 'info',
        format: winston.format.json(),
        transports
    });

    if (process.env.NODE_ENV !== 'production') {
        winstonLogger.add(new winston.transports.Console({
            format: winston.format.simple()
        }));
    }

    log = winstonLogger;
}


index.ts
// Import .env file if available
import * as dotenv from "dotenv";
dotenv.config({ path: __dirname+'/.env' });

import { ExternalTaskClient } from "./client";
import { log, creatLogger } from './log';

function main () {
    creatLogger(); // initialize logger
    log.info("Start micro API");

    const extClient = new ExternalTaskClient();
    const instance = extClient.init().getInstance();

    const topic = process.env.TOPIC;

    if (topic) {
        instance.subscribe(topic, taskSubscription);
        instance.start();
    }
}

main();

/***
 * Main task for subscription
 * @param task
 * @param taskService
 * @returns {Promise<void>}
 */
export async function taskSubscription({ task, taskService }){
    try {
        // get module which shall be executed
        const microModule = task.variables.get("micro-api-module");
        log.info(microModule);
        const module = await import('./modules/' + microModule);

        module.main({ task, taskService });
    }
    catch (ex) {
         // log.error(ex);
        await taskService.handleFailure(task, "An error occured - strange ...");
    }
}


modules/test.http-request.ts
import { Variables } from 'camunda-external-task-client-js';
import * as request from 'postman-request';

/***
 * Main task for subscription
 * @param task
 * @param taskService
 * @returns {Promise<void>}
 */
export async function main({ task, taskService }){
    try{
        const processVariables  = new Variables();

        try{
            let ping = task.variables.get("ping") || undefined;
            console.log(`var ping: ${ping}`);

            // Demo response
            /*
            {
                "userId": 1,
                "id": 1,
                "title": "delectus aut autem",
                "completed": false
            }
            */
            request({
                url: 'https://jsonplaceholder.typicode.com/todos/1',
                method: 'get'
            }, async function (error, response, body) {
                console.log('error:', error); // Print the error if one occurred
                console.log('statusCode:', response && response.statusCode); // Print the response status code if a response was received

                const content = JSON.parse(body);

                // set something from response as process variable
                processVariables.set("title", content.title);

                await taskService.complete(task, processVariables);
            });
        } catch(ex){
            console.error(ex);
            await taskService.handleFailure(task, "An error occurred while requesting ressource");
        }
    }
    catch(ex){
        // log.error(ex);
        await taskService.handleFailure(task, "An error occured - strange ...");
    }
};


client.ts
import { Client, BasicAuthInterceptor } from "camunda-external-task-client-js";
import { util } from "camunda-external-task-client-js/lib/__internal/utils";
import { log as log } from './log';

export class ExternalTaskClient {
    instance: ClientInstance;
    constructor(options?: object) {
        let env = {};

        // If options are available use them
        if (options) {
            env = options
        }

        process.env = Object.assign(process.env, env);
    }

    init(options?: options) {
        options = options || {};
        let interceptors,
            config = {},
            use;

        let maxTasks =      this.parseEnv('MAX_TASK', 'integer'),
            interval =      this.parseEnv('INTERVAL', 'integer', 300),
            lockDuration =  this.parseEnv('LOCK_DURATION', 'integer', 50000),
            autoPoll =      this.parseEnv('AUTO_POLL', 'boolean', false),
            workerId =      this.parseEnv('WORKER_ID', 'string', 'default-worker'),
            baseUrl =       this.parseEnv('BASE_URL', 'string', 'http://localhost:8080/engine-rest'),
            AUTH =          this.parseEnv('AUTH', 'boolean', false),
            asyncResponseTimeout = this.parseEnv('ASYNC_RESPONSE_TIMEOUT', 'integer');

        //Use basic auth if needed for Camunda rest engine
        if(AUTH === true){
            let basicAuth = new BasicAuthInterceptor({
                username: this.parseEnv('USER', 'string', 'demo'),
                password: this.parseEnv('PASS', 'string', 'demo')
            });

            if (!interceptors) interceptors = [];

            interceptors.push(basicAuth);
        }

        //Add optional middleware for use
        if(
            (options.use && typeof options.use === 'function') ||
            (options.use && util.isArrayOfFunctions(options.use))
        ){
            use = options.use;
        }

        config = {
            maxTasks,
            asyncResponseTimeout,
            interval,
            lockDuration,
            autoPoll,
            interceptors,
            use,
            workerId,
            baseUrl
        };

        this.instance = new Client(config);
        this.appendLogger();

        return this;
    }

    parseEnv(envVar: string, typeOf: string, defaults: string|number|boolean = undefined) : string|number|boolean {
        let envValue = process.env[envVar];

        if (typeOf === 'integer') {
            let envValInt = Number.parseInt(envValue);
            return Number.isNaN(envValInt) ? defaults : envValInt;
        }
        else if (typeOf === 'boolean') {
            let envValBool = (envValue == undefined ? defaults : envValue); // check if default is available instead
            return (envValBool === 'true' || envValBool === true); // could be default bool or envVal
        }
        else {
            return envValue ? envValue : defaults;
        }
    }

    appendLogger () {
        this.instance.on("subscribe", topic => {
            log.info('Subscribed ' + topic);
        });

        this.instance.on("unsubscribe", topic => {
            log.info('Unsubscribed ' + topic);
        });

        this.instance.on("poll:start", () => {
            log.info('polling');
        });

        this.instance.on("poll:stop", () => {
            log.error("X-Camunda-PollError: stopped");
        });

        this.instance.on("poll:success", tasks => {
            log.info('task ' + tasks);
        });

        this.instance.on("poll:error", e => {
            log.error('X-Camunda-PollError: ' + e);
        });

        this.instance.on("complete:success", ({ id }) => {
            log.info(`completed task ${id}`);
        });

        this.instance.on("complete:error", ({ id }, e) => {
            log.error(`X-Camunda-CompleteError: couldn't complete task ${id}, ${e}`);
        });

        this.instance.on("handleFailure:success", ({ id }) => {
            log.info(`handled failure of task ${id}`);
        });

        this.instance.on("handleFailure:error", ({ id }, e) => {
            log.error(`X-Camunda-FailureError: couldn't handle failure of task ${id}, ${e}`);
        });

        this.instance.on("handleBpmnError:success", ({ id }) => {
            log.info(`handled BPMN error of task ${id}`);
        });

        this.instance.on("handleBpmnError:error", ({ id }, e) => {
            log.error(`X-Camunda-BpmnError: couldn't handle BPMN error of task ${id}, ${e}`);
        });

        this.instance.on("extendLock:success", ({ id }) => {
            log.info(`handled extend lock of task ${id}`);
        });

        this.instance.on("extendLock:error", ({ id }, e) => {
            log.error(`X-Camunda-ExtendLockError: couldn't handle extend lock of task ${id}, ${e}`);
        });

        this.instance.on("unlock:success", ({ id }) => {
            log.info(`unlocked task ${id}`);
        });

        this.instance.on("unlock:error", ({ id }, e) => {
            log.error(`X-Camunda-UnlockError: couldn't unlock task ${id}, ${e}`);
        });
    }

    getInstance () {
        return this.instance;
    }
}

export interface ExternalTaskClient {
    instance: ClientInstance
    init: (options?: options) => this
    parseEnv: (envVar: string, typeOf: string, defaults?: string|number|boolean) => string|number|boolean
    getInstance: () => ClientInstance
}

interface options {
    interceptors?: () => any | []
    use?: () => any | []
}

interface ClientInstance {
    subscribe: (topic: string, module: any) => any
    start: () => any
    on: (name: string, fn: (param1?: any, param2?: any) => void) => any
}
